### PR TITLE
Add nav2 rrtconnect planner plugin

### PR DESCRIPTION
**Pull request by shivaang12**
* Add ompl rrt connect planner

### CODE REVIEW

File: nav2_rrtconnect_planner/include/nav2_rrtconnect_planner/rrt_connect.hpp [hunk 1]
```diff
@@ -0,0 +1,130 @@
+/*********************************************************************
+ *
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2008, 2013, Willow Garage, Inc.
```
Comment (SteveMacenski) [inline nav2_rrtconnect_planner/include/nav2_rrtconnect_planner/rrt_connect.hpp:5]: Is this willow code?

File: nav2_rrtconnect_planner/include/nav2_rrtconnect_planner/rrt_connect.hpp [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Shivang Patel
+ *
+ * Reference tutorial:
+ * https://ros-planning.github.io/navigation2/tutorials/docs/writing_new_nav2planner_plugin.html
+ *********************************************************************/
+#ifndef NAV2_RRTCONNECT_PLANNER_INCLUDE_NAV2_RRTCONNECT_PLANNER_RRT_CONNECT_HPP
```
Comment (SteveMacenski) [inline nav2_rrtconnect_planner/include/nav2_rrtconnect_planner/rrt_connect.hpp:41]: This guard is incorrect - please run the ament_cpplint ament_xmllint ament_uncrustify

File: nav2_rrtconnect_planner/include/nav2_rrtconnect_planner/rrt_connect.hpp [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+#define NAV2_RRTCONNECT_PLANNER_INCLUDE_NAV2_RRTCONNECT_PLANNER_RRT_CONNECT_HPP
+
+#include <ompl/base/spaces/RealVectorStateSpace.h>
```
Comment (SteveMacenski) [inline nav2_rrtconnect_planner/include/nav2_rrtconnect_planner/rrt_connect.hpp:44]: Use quotes please

File: nav2_rrtconnect_planner/package.xml [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  <depend>visualization_msgs</depend>
+  <depend>nav2_util</depend>
+  <depend>nav2_msgs</depend>
+  <depend>nav_msgs</depend>
+  <depend>geometry_msgs</depend>
+  <depend>builtin_interfaces</depend>
+  <depend>tf2_ros</depend>
+  <depend>nav2_costmap_2d</depend>
+  <depend>nav2_core</depend>
+  <depend>pluginlib</depend>
```
Comment (SteveMacenski) [inline nav2_rrtconnect_planner/package.xml:25]: You need an OMPL key

File: nav2_rrtconnect_planner/src/rrt_connect.cpp [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Shivang Patel
+ *
+ * Reference tutorial:
+ * https://ros-planning.github.io/navigation2/tutorials/docs/writing_new_nav2planner_plugin.html
+ *********************************************************************/
+#include <ompl/base/spaces/SE2StateSpace.h>
```
Comment (SteveMacenski) [inline nav2_rrtconnect_planner/src/rrt_connect.cpp:41]: Quotes and move after the STL headers (iostream...utility)

File: nav2_rrtconnect_planner/src/rrt_connect.cpp [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+  RCLCPP_INFO(
+    node_->get_logger(), "Deactivating plugin %s of type NavfnPlanner",
+    name_.c_str());
+}
+
+void RRTConnect::initializeOMPL()
+{
+  // Defining Bounds
+  bounds_ = new ompl::base::RealVectorBounds(2);
+  bounds_->setLow(0, costmap_->getOriginX());
```
Comment (SteveMacenski) [inline nav2_rrtconnect_planner/src/rrt_connect.cpp:124]: My recollection is that you typically set some normalized bounds and transform your configuration space into it afterwards.

File: nav2_rrtconnect_planner/src/rrt_connect.cpp [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+    costmap_->getOriginY());
+
+  RCLCPP_INFO(
+    node_->get_logger(),
+    "bounds: %f %f %f %f", bounds_->low[0], bounds_->low[1],
+    bounds_->high[0], bounds_->high[1]);
+
+  ompl_state_space_ = std::make_shared<ompl::base::SE2StateSpace>();
+  // Bounding the R2 state space
+  ompl_state_space_->as<ompl::base::SE2StateSpace>()->setBounds(*bounds_);
```
Comment (SteveMacenski) [inline nav2_rrtconnect_planner/src/rrt_connect.cpp:140]: If you dereference `bounds_` you don't need it as a pointer.

File: nav2_rrtconnect_planner/src/rrt_connect.cpp [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+  ss_->getSpaceInformation()->setStateValidityCheckingResolution(collision_checking_resolution_);
+
+  ompl_initialized_ = true;
+}
+
+nav_msgs::msg::Path RRTConnect::createPlan(
+  const geometry_msgs::msg::PoseStamped & start,
+  const geometry_msgs::msg::PoseStamped & goal)
+{
+  if (!ompl_initialized_) {
```
Comment (SteveMacenski) [inline nav2_rrtconnect_planner/src/rrt_connect.cpp:159]: You should update this if the map dimensions change

File: nav2_rrtconnect_planner/src/rrt_connect.cpp [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+  ss_->setPlanner(
+    ompl::base::PlannerPtr(
+      new ompl::geometric::RRTConnect(
+        ss_->getSpaceInformation())));
+  // Load settings
+  ss_->setup();
+  // Counter for replanning
+  int counter = 0;
+  // Planning
+  while (!ss_->solve(solve_time_) && counter < max_trials_) {
```
Comment (SteveMacenski) [inline nav2_rrtconnect_planner/src/rrt_connect.cpp:196]: Why is this necessary? Does it fail that often?

File: nav2_rrtconnect_planner/src/rrt_connect.cpp [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+    global_path.poses.clear();
+    global_path.header.stamp = node_->now();
+    global_path.header.frame_id = global_frame_;
+    // Increasing number of path points
+    int min_num_states = round(solution_path.length() / costmap_->getResolution());
+    solution_path.interpolate(min_num_states);
+
+    global_path.poses.reserve(solution_path.getStates().size());
+    for (const auto ptr : solution_path.getStates()) {
+      global_path.poses.emplace_back(getPoseStampedFromState(*ptr));
```
Comment (SteveMacenski) [inline nav2_rrtconnect_planner/src/rrt_connect.cpp:217]: If you're constructing the object, you don't get a speed up on emplace back.

File: nav2_rrtconnect_planner/src/rrt_connect.cpp [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+
+  // Simple setup
+  ss_.reset(new ompl::geometric::SimpleSetup(ompl_state_space_));
+  // Setting state validity checker
+  ss_->setStateValidityChecker(
+    [this](const ompl::base::State * state) {
+      return this->isStateValid(state);
+    });
+  // Setting the state validity checking resolution
+  ss_->getSpaceInformation()->setStateValidityCheckingResolution(collision_checking_resolution_);
```
Comment (SteveMacenski) [inline nav2_rrtconnect_planner/src/rrt_connect.cpp:150]: `collision_checking_resolution_`, I'm not sure off hand what that means - can this be the costmap resolution?

File: nav2_rrtconnect_planner/include/nav2_rrtconnect_planner/rrt_connect.hpp (hunk 1 above)
Comment (shivaang12) [reply nav2_rrtconnect_planner/include/nav2_rrtconnect_planner/rrt_connect.hpp:5]: I thought about the ROS2 platform. I don't have much experience with licensing so please guide me.

File: nav2_rrtconnect_planner/include/nav2_rrtconnect_planner/rrt_connect.hpp (hunk 2 above)
Comment (shivaang12) [reply nav2_rrtconnect_planner/include/nav2_rrtconnect_planner/rrt_connect.hpp:41]: cool!

File: nav2_rrtconnect_planner/include/nav2_rrtconnect_planner/rrt_connect.hpp (hunk 3 above)
Comment (shivaang12) [reply nav2_rrtconnect_planner/include/nav2_rrtconnect_planner/rrt_connect.hpp:44]: okay!

File: nav2_rrtconnect_planner/src/rrt_connect.cpp (hunk 6 above)
Comment (shivaang12) [reply nav2_rrtconnect_planner/src/rrt_connect.cpp:124]: okay. Let me try this approach.

File: nav2_rrtconnect_planner/src/rrt_connect.cpp (hunk 9 above)
Comment (shivaang12) [reply nav2_rrtconnect_planner/src/rrt_connect.cpp:196]: Not that often but I think it might help if it fails in first try.

File: nav2_rrtconnect_planner/src/rrt_connect.cpp (hunk 10 above)
Comment (shivaang12) [reply nav2_rrtconnect_planner/src/rrt_connect.cpp:217]: Got it!

File: nav2_rrtconnect_planner/src/rrt_connect.cpp (hunk 11 above)
Comment (shivaang12) [reply nav2_rrtconnect_planner/src/rrt_connect.cpp:150]: This resolution is for the state validator. It means how deep the validator has to go if checking for obstacles between two sampled states. If you pass costmap resolution it doesn't work nice.

**Comment by shivaang12** [comment]
> What's the speed of this? For instance, across the map in https://github.com/turtlebot/turtlebot_apps/tree/indigo/turtlebot_navigation/maps. Can you post some example paths for me to visualize?

I do have a youtube video if that helps!
https://youtu.be/MSy7WAE_xz4

**Comment by shivaang12** [comment]
> A value point in sampling based is to have state - you should use a costmap model and check the validity of the footprint so it works with non-circular robots.
A value prop on OMPL is using many planners easily (https://github.com/windelbouwman/move-base-ompl/blob/master/src/ompl_global_planner.cpp#L308) you should parameterize some enum to expose all to a user. Its a 5 minute task that could be really helpful if RRT connect isnt the right solution.

I thing this would be a great addition. Actually I was planning on doing these after this PR gets through. But I have no problem of doing it now.

File: nav2_rrtconnect_planner/include/nav2_rrtconnect_planner/rrt_connect.hpp (hunk 1 above)
Comment (SteveMacenski) [reply nav2_rrtconnect_planner/include/nav2_rrtconnect_planner/rrt_connect.hpp:5]: If you copy pasted code from somewhere, you should add their license. If you made this yourself, then you don't.

File: nav2_rrtconnect_planner/src/rrt_connect.cpp (hunk 11 above)
Comment (SteveMacenski) [reply nav2_rrtconnect_planner/src/rrt_connect.cpp:150]: If 0.05 is "typical", I'd think 0.025 to be reasonable (2 sample potentially per pixel). Or am I missing something?

**Comment by SteveMacenski** [comment]
If you make a chrono timer (https://learntechway.com/simple-timer-in-c-using-chrono-library/) on the start and end of your callback you can get actual metrics on speed

File: nav2_rrtconnect_planner/src/rrt_connect.cpp (hunk 11 above)
Comment (shivaang12) [reply nav2_rrtconnect_planner/src/rrt_connect.cpp:150]: 0.025 also doesn't work. But 0.0125 does work. I am not sure but I think it some thing related to interpolation which `ompl::base::DiscreteMotionValidator` uses for checking obstacles between two states.

Comment (SteveMacenski) [reply nav2_rrtconnect_planner/src/rrt_connect.cpp:150]: Mhm, maybe something to we should try to understand more. "How" didnt it work? Do you know what the units for this are? Is it configuration space units (so for now the costmap size, so meters I believe from yoru setup)? Or interpolated normalized? 

As you can tell, its been awhile since I looked at OMPL code, apologies. I should know this better.

Comment (shivaang12) [reply nav2_rrtconnect_planner/src/rrt_connect.cpp:150]: No worries! It's okay not to know :smiley:. This is my first time working with OMPL. I would dig deeper and try to find out about the problem.

**Comment by shivaang12** [comment]
That would be 
mean: 8481.04
max: 23855.0
min: 1922.0
all in microseconds

**Comment by SteveMacenski** [comment]
Ok, that's a pretty short path which isn't really indicative of much. My recommendation is to use that willow map I linked you to. In the launch files, you can change the TB3 world to that. As a hack, you can remove the obstacle/voxel layers so you have only the static and inflation layers. Then you can launch everything and test a few paths (though the robot will try to move through the TB3 world and eventually crash or something, so only good for quick tests. I've been doing that this week for some quick prototyping on planning).

File: nav2_rrtconnect_planner/src/rrt_connect.cpp [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+    bounds_->high[0], bounds_->high[1]);
+
+  ompl_state_space_ = std::make_shared<ompl::base::SE2StateSpace>();
+  // Bounding the R2 state space
+  ompl_state_space_->as<ompl::base::SE2StateSpace>()->setBounds(*bounds_);
+
+  // Simple setup
+  ss_.reset(new ompl::geometric::SimpleSetup(ompl_state_space_));
+  // Setting state validity checker
+  ss_->setStateValidityChecker(
```
Comment (SteveMacenski) [inline nav2_rrtconnect_planner/src/rrt_connect.cpp:145]: Oh this is odd, why is this a lambda? Does a typical std::bind not work?

**Comment by shivaang12** [comment]
Lets get over with tutorial first. I will removed OMPL stuff from the tutorial PR.

File: nav2_rrtconnect_planner/src/rrt_connect.cpp (hunk 12 above)
Comment (shivaang12) [reply nav2_rrtconnect_planner/src/rrt_connect.cpp:145]: std::bind does work but I thought lambda would be more efficient. what do you think?

**Comment by shivaang12** [comment]
Data for the willograge map
For Medium goal:
mean: 6029.151898734177
max: 54937.0
min: 632.0

For farthest goal
mean: 10043.780487804877
max: 23019.0
min: 301.0

All time in microseconds

File: nav2_rrtconnect_planner/include/nav2_rrtconnect_planner/rrt_connect.hpp [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+#include "nav_msgs/msg/path.hpp"
+#include "nav2_util/robot_utils.hpp"
+#include "nav2_util/lifecycle_node.hpp"
+#include "nav2_costmap_2d/costmap_2d_ros.hpp"
+
+
+namespace nav2_rrtconnect_planner
+{
+
+class RRTConnect : public nav2_core::GlobalPlanner
```
Comment (mamoll) [inline nav2_rrtconnect_planner/include/nav2_rrtconnect_planner/rrt_connect.hpp:63]: There is very little that is RRTConnect specific in this class. I would name it OMPLPlanner (or something like that). You can make RRTConnect the default planner, if that seems to work best. You should add a method to set the Planner. Perhaps something like this:
```
template <typename Planner>
void setPlanner()
{
    if (!ss_)
       initializeOMPL();
   ss_->setPlanner(std::make_shared<Planner>(ss_->getSpaceInformation());
}
```

Since you are using cost maps, wouldn't you want to use a planner that is cost-aware? To do so, you first need to create an [ompl::base::OptimizationObjective](http://ompl.kavrakilab.org/classompl_1_1base_1_1OptimizationObjective.html)-derived class that captures the state and/or motion cost of moving over this cost map. Once you have that, try running TRRT or BiTRRT. PRM*/RRT* might also be reasonable choices.

If the path cost is simply distance traveled, then there's no need to set up an OptimizationObjective. This is the default cost that is being optimized when using a cost-aware planner like TRRT, BiTRRT, PRM*, RRT*, ...

File: nav2_rrtconnect_planner/src/rrt_connect.cpp (hunk 12 above)
Comment (SteveMacenski) [reply nav2_rrtconnect_planner/src/rrt_connect.cpp:145]: std::bind doesn't require creating a new lambda to just call the same function - we should just use that.

Comment (mamoll) [reply nav2_rrtconnect_planner/src/rrt_connect.cpp:145]: Scott Meyers in his book Effective Modern C++ makes a good case for using lambda exprs instead of std::bind. See https://github.com/peter-can-write/cpp-notes/blob/master/prefer-lambdas-over-std::bind.md

Comment (SteveMacenski) [reply nav2_rrtconnect_planner/src/rrt_connect.cpp:145]: I disagree with the initial premise `Lambdas are more readable and clear...` but I see the point. This is fine as is.

File: nav2_rrtconnect_planner/src/rrt_connect.cpp [hunk 14]
```diff
@@ -0,0 +1,269 @@
...
+  ompl::base::ScopedState<> ss(ompl_state_space_);
+  ss = state;
+  unsigned int mx, my;
+  costmap_->worldToMap(ss[0], ss[1], mx, my);
+
+  auto cost = costmap_->getCost(mx, my);
+  if (cost == nav2_costmap_2d::LETHAL_OBSTACLE ||
+    cost == nav2_costmap_2d::INSCRIBED_INFLATED_OBSTACLE ||
+    ( !allow_unknown_ && cost == nav2_costmap_2d::NO_INFORMATION))
+  {
```
Comment (ardabbour) [inline nav2_rrtconnect_planner/src/rrt_connect.cpp:240]: With the way this state validation check is done, it seems that using an SE2 space adds an extra rotational dimension that will be searched in without any benefit

File: nav2_rrtconnect_planner/src/rrt_connect.cpp [hunk 15]
```diff
@@ -0,0 +1,269 @@
...
+  while (!ss_->solve(solve_time_) && counter < max_trials_) {
+    ++counter;
+  }
+  // Plan found
+  if (counter < max_trials_) {
+    RCLCPP_INFO(
+      node_->get_logger(), "Found %d solutions",
+      ss_->getProblemDefinition()->getSolutionCount());
+    // Path simplification
+    ss_->simplifySolution();
```
Comment (ardabbour) [inline nav2_rrtconnect_planner/src/rrt_connect.cpp:205]: This maybe out of the scope of a simple tutorial but choosing between returning a simplified/original solution could be a parameter.

File: nav2_rrtconnect_planner/src/rrt_connect.cpp [hunk 16]
```diff
@@ -0,0 +1,269 @@
...
+ *
+ * Author: Shivang Patel
+ *
+ * Reference tutorial:
+ * https://ros-planning.github.io/navigation2/tutorials/docs/writing_new_nav2planner_plugin.html
+ *********************************************************************/
+#include <ompl/base/spaces/SE2StateSpace.h>
+#include <ompl/geometric/planners/rrt/RRTConnect.h>
+#include <ompl/util/PPM.h>
+#include <ompl/tools/config/MagicConstants.h>
```
Comment (ardabbour) [inline nav2_rrtconnect_planner/src/rrt_connect.cpp:44]: Are any PPM files or magic constants used?

Comment (shivaang12) [reply nav2_rrtconnect_planner/src/rrt_connect.cpp:44]: My bad!

**Comment by SteveMacenski** [comment]
@shivaang12 maybe take into account these comments and then submit a new draft PR with those changes and any others you want to make on the main github page

**Comment by SteveMacenski** [comment]
Closing since no movement in awhile and we made a simpler path planning tutorial. OMPL work belongs in main repo once we get back into it.
